File: medium/491.go
package main

var sol, nums []int
var ret [][]int
var sn int

func search(p, v1 int) {
  if len(sol) >= 2 {
    cpy := make([]int, len(sol))
    copy(cpy, sol)
    ret = append(ret, cpy)
  }
  s := len(sol)
  for i := p; i < sn; i++ {
    v := nums[i]
    if v <= v1 {
      continue
    }
    before := false
    for j := p; j < i; j++ {
      if nums[j] == v {
        before = true
        break
      }
    }
    if before {
      continue;
    }
    for j := i; j < sn; j++ {
      if nums[j] == v {
        sol = append(sol, v)
        search(j + 1, v)
      }
    }
    sol = sol[:s]
  }
}

func findSubsequences(nums_ []int) [][]int {
  nums = nums_
  sn = len(nums)
  sol = []int{}
  ret = [][]int{}
  search(0, -101)
  return ret
}

func main() {
  nums := [...]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15}
  for i := 0; i < 100; i++ {
    findSubsequences(nums[:])
  }
}

File: leetcode/medium/491.cpp
#include <vector>
#include <numeric>

class Solution {
  std::vector<int> sol;
  std::vector<std::vector<int>> ret;
  void search(std::vector<int> & nums, int p, int v1) {
    int sn = nums.size();
    if (sol.size() >= 2) {
      ret.push_back(sol);
    }
    int s = sol.size();
    for (int i = p; i < sn; ++i) {
      int v = nums[i];
      if (v <= v1) {
        continue;
      }
      bool before = false;
      for (int j = p; j < i; ++j) {
        if (nums[j] == v) {
          before = true;
          break;
        }
      }
      if (before) {
        continue;
      }
      for (int j = i; j < sn; ++j) {
        if (nums[j] == v) {
          sol.push_back(v);
          search(nums, j + 1, v);
        }
      }
      sol.resize(s);
    }
  }
public:
  std::vector<std::vector<int>> findSubsequences(std::vector<int>& nums) {
    search(nums, 0, -101);
    return ret;
  }
};

int main() {
  std::vector<int> input(15);
  std::iota(input.begin(), input.end(), 1);
  for (int i = 0; i < 100; ++i) {
    Solution sol;
    sol.findSubsequences(input);
  }
}
